### DISCUSSION TITLE
Wavelog hardware interface proof of concept

### DISCUSSION THREAD

**Discussion by filipmelik** [wavelog/wavelog#1316] [Show and tell]
Hi! 
I am happy user of the wavelog and I was using the radio frequency CAT reading through my PC. 
However, I find it kind of cumbersome having need for my PC being open and running and also need for having two pieces of software (wlgate +flrig) running on it to get my frequency reading sent to wavelog, when doing just some ad-hoc contacts using wavelog on my phone.

So I was tinkering with this idea of having external hardware interface that will send my current frequency reading into the wavelog without any  big and cumbersome computer connected. 

Here I present my first experiment which has a few rough edges, but it works with my IC-7300 rig so far.

It is basic ESP32 development board WROOM32 with external display for convenience and its connected to 7300’s CI-V port.

I have some idea to make this more general so other rigs can be used too, but that would need more work to be done and I am not promising anything now. so Just wanted to show what I have for now 😁

https://youtu.be/5TSYUnd2EUc?si=3oXh6kyRyE0OsNX8

**Comment by HB9HIL** [comment]
Hello @filipmelik 

You are not the first one 😃 

There is already a CI-V solution here: https://github.com/dg9vh/Wavelog_CI_V by @dg9vh 
and a solution for Yaesu and other radios here: https://github.com/zone11/WLBridge4TRX by @zone11

**Reply by HB9HIL** [reply]
https://github.com/wavelog/Wavelog/wiki/Third-Party-Tools

**Reply by dg9vh** [reply]
But a nice solution with display! So just go on like this! We are not concurrence, we enrich each other :-)

**Reply by filipmelik** [reply]
Nice! Why this was not available last time I was looking for this? 😁 anyways, what is the fun in it if I did not program it myself? 😂 will definitely check the existing solutions and possibly get some inspiration 💪

**Reply by dg9vh** [reply]
Is your code available? So we can learn from each other :-) I know my solution isn't the best in the world :-) And I allways want to learn how others did things I did a complicated way .-)

**Comment by zone11** [comment]
Well done @filipmelik, I love the idea of the small display.
Maybe you could share your code to allow @dg9vh, myself and maybe others to learn something. 
My solution is pragmatically cobbled together and I am curious to see how you have implemented the timing for querying the TRX.

**Reply by filipmelik** [reply]
Hey Christian and Kim! Thanks for sharing your code on github, as I said, I will try to go over it in the evening to see how your code works in more detail, because I am as curious as you two are 😂

anyways, there is one significant difference, and that is that I used micropython to program my proof of concept as I find it faster to work in. This is honestly my first project with ESP32 and micropython (I used only arduinos in the past, but for nothing “complicated” as this).

there is currently not much to share, as my code is hacked together in very nasty way, just a proof of concept as I said. However, there are some ideas I am carrying in my head on how to make this thing great. I will try to write it down later today and see what you think, maybe we will spark some collab or something 🫣💪

**Reply by filipmelik** [reply]
So, just to reiterate why I am doing this and what is my use case:

I work on home office 99% of the time and my radio is sitting next to me. Of course I am working very hard 😂 And since I have [HamAlert](https://hamalert.org/) configured, occasionally there is a station of the interest on the air I want to make QSO with, or I just have few spare minutes over lunch break to make some random QSOs. So I open the wavelog, because up until now I am logging every QSO. Because I am lazy person, I really hate that I need to input the frequency band and mode manually into to wavelog for every random QSO, so I hooked up my rig to my computer. But in order to make frequency reading work, I also need to turn on the WLGate and flrig and also hope, that the usb-to-serial converter works, because often times, I need to reconnect the usb cable anyway for some unknown reasons to make it work again. I got fed up with this pretty quickly, co I was thinking that I need something to automate this (I love automatic things, that's why I love wavelog, it uploads QSOs for you automatically and all that jazz 😁). And that is why this idea, which obviously I am not the first one who had it, was born.

 My vision with this piece of hardware is as follows:
- It should have display to be able to quickly check what is the state of the device
- The display should display wifi signal strength, frequency, band, maybe something else
- It should work automatically without any interaction of the user (Of course, apart from initial configuration). This means, that it should ideally power on automatically with the radio and turn off when radio is off
- The configuration should be done the same way how @dg9vh has implemented it: There should be wifi server I connect to, and configure the parameters there (URL, API key, WIFI SSID, WIFI PASS, RIG type, polling interval). 
- This "settings" mode should be invoked by user pressing some button. When pressing the button, the display would display the WIFI AP SSID and IP address, where the HTTP server is running, So I do not have to fish for it using serial console or similar. This allows easy reconfiguration in case it's needed. 
- The device should read and the frequency + band (and power, if supported) data periodically using the polling mechanism. The Icom has this "broadcast when parameters changes" function, but that is just an Icom thing as far as I know. 
- Ideally, it should be universal, I mean rig-wise. The ideal how to achieve this is that the device should be able to upload "rig driver" using the configuration page. Where to get this rig driver? Well, there are few softwares, that already use some drivers for their use. Flrig use configuration XML files (This is called RigCat afaik), Omnirig is using ini files, hamlib uses "proprietary" drivers written in C++, ther eare of course some other options I am probably not aware of. My idea, for ease of use, is that we should use the omnirig format for this. Why? because it is proven to work, well documented, it is text-based, so it is easy to read and write, and should be "easiest" in terms of writing the parser for ESP32. I feel that XML parsing would be much heavier using micropython (not confirmed, just my feeling). This would enable this device to be "universal" on the software side of rig control.
- The hardware design should be able to support the different rigs hardware-wise (all radios have different connectors, maybe voltage levels, etc.)
- It would be cool if we also had "production grade HW" ready, to make it as small and bulletproof as possible. Anyway, I am not good at this, but somebody could design PCB as an open hardware, that could be then ordered using [JLCPCB](https://jlcpcb.com/) or similar service. And the 3D printed case is the similar story here.

Maybe I forgot something while writing this, so I might get back to this list later.

What do you think? Any other ideas? What I have described is an "ideal" state, and for sure this will be a lot of work. I am not promising anything here, just wanted to share my vision for now 👍 

but to be honest I do not even know if there is a demand for such thing, or if this just beneficial for 3 persons, us 😂

**Reply by zone11** [reply]
Looks like you have the same intentions like some of us. I absolutely identify myself with your vision.
Except the need of a display, everything else was on my wish list, but I had to resize a bit due the lack of time and a solution worked fine in the past half year.

I like the idea of having one hardware design for serial CAT (different levels) and CI-V, but also multiple options to power the board:
- Internal Small LiPo with RS232 like the an Elecraft KX in a setup with very limited space
- 12V from the TRX ACC with RS232 Serial like the FT-991
- 12V from the TRX with TTL Level Serial like the FT-710 

The use of the informations from the Omnirig files is exactly what's needed to send and parse the serial CAT data. I'm having a draft version of my code exactly working with this informations, but entered on the UI. Having an option to import a file or embed the required data in the code, having them parsed from the available Omnirig files, should not be such a big deal.
Thinking about the option to select the most used brands form a dropdown and having an option to enter custom informations. Also some debugging output is needed then.

I'm going to spend some time for the project during x-mas holidays:
- Have a look at alternative options for the currently used WifiManager
- Add a small OLED to display vital information (as an option) - found a neat ESP32board with an OLED on it.
- Add a RBG LED to show current status (as an option) - for very small setups without the need of a display
- Redo my CAT parsing and device selection
- Make the code more versatile to add the option for CI-V (I don't own ICOM hardware to develop it myself)

If you ore somebody else would like to discuss or challenge the ideas, I'm open for having a chat.

**Reply by filipmelik** [reply]
I have been tinkering with the software part for few last days. So far I am able to upload the “driver file” to device using the settings form (surprisingly, I was not able to find any micropython library for http server that supports parsing multipart data, so I had to port some python lib to micropython). Now I am onto parsing the driver for actual commands which will be fun. Especially when some radios have binary commands, other have ascii commands and since wavelog suports logging different frequencies due to possible active split mode, this will be fun to handle, apart from other possible edge cases not yet discovered 😂

I was thinking that when the code is finished and tested, we should, apart from source code itself, provide the “compiled” binary, for ease of installation by users, not to having to deal with installing various tools and IDEs to upload the code onto device. There should be just some bin uploader, like esptool or similar involved I think.

another idea I haf was about actually embedding all the “Minified” radio drivers into the actual device, so you could select them from dropdown, but I ditched this idea, becase theoretically every time new druver is added, we would need to rebuild the binary, and I am also not sure space-wise, as the esp32 has of course limited flash.

anyways, I am olaying with the thing in the evenings and so far it is jist ball of spaghetti code, but I am slowly making the progress 😅 I will leave the hardware part open until the software is somehow finished :)

**Reply by filipmelik** [reply]
Hello all, I was able to do some progress on this one. I have (somehow) working proof of concept that is able to chew up omnirig config files. I am currently using WROOM32 board and code is written in micropython. Current version is able to:

1] Start a HTTP server and present webserver for configuring the device
2] The radio configuration is determined by the omnirig driver file that you upload via the settings web page
3] Read and display the frequency from the rig. It can also handle the split operation, if the rig supports it (different RX and TX freqs) 
4] **Sometimes**, it also works to report the frequency data to the wavelog API (lol) - see below 

What does not work:
1] Reporting TX power. Omnirig unfortunately does not provide any command to read this value.
2] With the WROOM32 board, the API call to report frequency fails 90% of the time **in case the wavelog instance run on HTTPS**.  After many hours of investigation, it seems that the problem is that the WROOM32 board has very limited RAM size (even after many optimization I tried it does not work reliably). HTTPS TLS handshake is apparently too heavy for this little guy. However, there seems to be another ESP32 board called WROVER, which should have much more RAM available. I will order it and test the code with that.

What needs to be done:
1] Hardware (PCB), that will support many different serial bus voltage levels to be able to accomodate most of the rigs. I think that every user will choose which parts to assemble on PCB in respect of the rig he wants to use. So ICOM CI-V users, will have this schottky diode, RS232 users will assemble MAX232 (or similar) chip etc. Also the 
2] Compile the code once its finished, so the deployment is just uploading one hex/bin file to the ESP32 board
3] Write docs, instructions etc.
4] Test the omnirig parser with some radio that does use ASCII commands, not binary ones. I Have only ICOM IC-7300 that has binary commands those seems to work.
5] Different means of supplying the power
6] Something I surely forgot

I have created a little video presenting current capabilities. Not much of a quality, but just to get the idea. After first power up, it goes to settings mode. So I configure the thing and then you can see it in action - reading the frequency, and even the split capability. The "Api called" is mocked for now until I resolve the memory issue, but it server as indication that the request was sucessfully acknowledged by wavelog. The button serves as a means to go to settings mode. Sorry for the blur on the end, I noticed too late :) 

Video here:  [https://youtu.be/7-2nmZTughs](https://youtu.be/7-2nmZTughs)

I will work on this more once the other board arrives, so after new year most probably :) If you have any thoughts or feedback, it is very welcome!

Merry christmas!

**Reply by dg9vh** [reply]
Hi, it seems Micropython is not as efficient as C on an ESP32 - ok, my solution does not have as many functionality as your but API-connections to wavelog are working fine also with https... and there is no special library included for this to do.

But we are looking forward to see your work in the next weeks - it's looking very nice and I hope, you'll find a way to have https also working :-)

Merry xmas and hny.

**Reply by filipmelik** [reply]
Yes, micropython is surely not as efficient as C, however, I cannot imagine myself writing omnirig parsing code in C to be honest 😄  If I do not include the "extra functionality", the HTTPS also works for me, but after all this work, I do not want to give it all up, so hoping the board with more RAM will make this working and enable even more improvements :) But we will see 💪

**Reply by filipmelik** [reply]
Good news! ESP32-S3 N16R8 which has 8MB RAM instead of 100kB as the plain old regular WROOM solved the problem. Also the price for those boards are the same, which is weird, but very nice :D  The HTTPS calls to API now works reliably and I also added the ability to query via custom command I injected into the omni-rig config file, because omnirig itself does not support command for querying TX power. 

Next step is to restructure the code to be async, so I can also implement the XML-RPX server for cloudlog offline functionality. 
btw, the code so far is here, but it is still work in progress and subject to big change: https://github.com/filipmelik/wavelog-trx-interface

**Comment by filipmelik** [comment]
Hello all! 

I have finally compiled some first version of the docs. Wiki chapter about flashing firmware is yet to be done, but the rest of the instruction should be usable, although there is always room for improvment :) I know that both of you guys @dg9vh and @zone11  are probably happy with your hardware, so you have no intentions trying this one, but both projects of yours were very inspiring and I have stolen some of your ideas and implemented them into my device, hope you do not mind 😇 If interested, you can check it in the following link. I might even shot some video in future, when I will have my 3D printed case, so it looks better :)

https://github.com/filipmelik/wavelog-trx-interface
